src/Cli/dotnet/ReleasePropertyProjectLocator.cs

```diff
@@ -0,0 +1,215 @@
+        {
+            checkSolutions = shouldCheckSolutionsForProjects;
+        }
+
+        /// <param name="slnProjectPropertytoCheck">A property to enforce if we are looking into SLN files. If projects disagree on the property, throws exception.</param>
+        /// <returns>A project instance that will be targeted to publish/pack, etc. null if one does not exist.</returns>
+        public ProjectInstance GetTargetedProject(IEnumerable<string> slnOrProjectArgs, Dictionary<string, string> globalProps, string slnProjectPropertytoCheck = "")
+        {
+            string potentialProject = "";
+
+            foreach (string arg in slnOrProjectArgs.Append(Directory.GetCurrentDirectory()))
+            {
+                if (IsValidProjectFilePath(arg))
+                {
+                    return TryGetProjectInstance(arg, globalProps);
+                }
+                else if (Directory.Exists(arg)) // We should get here if the user did not provide a .proj or a .sln
+                {
+                    try
+                    {
+                        return TryGetProjectInstance(MsbuildProject.GetProjectFileFromDirectory(arg).FullName, globalProps);
+                    }
+                    catch (GracefulException)
+                    {
+                        // Fall back to looking for a solution if multiple project files are found.
+                        string potentialSln = Directory.GetFiles(arg, "*.sln", SearchOption.TopDirectoryOnly).FirstOrDefault();
+
+                        if (!string.IsNullOrEmpty(potentialSln))
+                            return GetSlnProject(potentialSln, globalProps, slnProjectPropertytoCheck);
+                    } // If nothing can be found: that's caught by MSBuild XMake::ProcessProjectSwitch -- don't change the behavior by failing here. 
+                }
+            }
+
+            return string.IsNullOrEmpty(potentialProject) ? null : TryGetProjectInstance(potentialProject, globalProps);
+        }
+
+        /// <returns>The executable project (first if multiple exist) in a SLN. Returns null if no executable project. Throws exception if two executable projects disagree
+        /// in the configuration property to check.</returns>
+        public ProjectInstance GetSlnProject(string slnPath, Dictionary<string, string> globalProps, string slnProjectConfigPropertytoCheck = "")
+        {
+            // This has a performance overhead so don't do this unless opted in.
+            if (!checkSolutions)
+                return null;
+            Debug.Assert(slnProjectConfigPropertytoCheck == MSBuildPropertyNames.PUBLISH_RELEASE || slnProjectConfigPropertytoCheck == MSBuildPropertyNames.PACK_RELEASE, "Only PackRelease or PublishRelease are currently expected");
+
+            SlnFile sln;
+            try
+            {
+                sln = SlnFileFactory.CreateFromFileOrDirectory(slnPath);
+            }
+            catch (GracefulException)
```

Comment:
Any chance the exception might hide important information when caught and not reported?

Reply:
I don't think so, because MSBuild will still be called, and it should be what handles the errors here.